// nexus/dist/definitions/interfaceType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var definitionBlocks_1 = require("./definitionBlocks");
var _types_1 = require("./_types");
var graphql_1 = require("graphql");
var InterfaceDefinitionBlock = /** @class */ (function (_super) {
    tslib_1.__extends(InterfaceDefinitionBlock, _super);
    function InterfaceDefinitionBlock(typeBuilder) {
        var _this = _super.call(this, typeBuilder) || this;
        _this.typeBuilder = typeBuilder;
        return _this;
    }
    /**
     * Sets the "resolveType" method for the current type.
     */
    InterfaceDefinitionBlock.prototype.resolveType = function (fn) {
        this.typeBuilder.setResolveType(fn);
    };
    return InterfaceDefinitionBlock;
}(definitionBlocks_1.OutputDefinitionBlock));
exports.InterfaceDefinitionBlock = InterfaceDefinitionBlock;
var NexusInterfaceTypeDef = /** @class */ (function () {
    function NexusInterfaceTypeDef(name, config) {
        this.name = name;
        this.config = config;
        graphql_1.assertValidName(name);
    }
    Object.defineProperty(NexusInterfaceTypeDef.prototype, "value", {
        get: function () {
            return this.config;
        },
        enumerable: true,
        configurable: true
    });
    return NexusInterfaceTypeDef;
}());
exports.NexusInterfaceTypeDef = NexusInterfaceTypeDef;
_types_1.withNexusSymbol(NexusInterfaceTypeDef, _types_1.NexusTypes.Interface);
/**
 * Defines a GraphQLInterfaceType
 * @param config
 */
function interfaceType(config) {
    return new NexusInterfaceTypeDef(config.name, config);
}
exports.interfaceType = interfaceType;
